### PR TITLE
Memory management raw pointers convertions

### PR DESCRIPTION
**Pull request by glevkovich**
* **Problem Overview**  
Few raw pointers to smart pointers code refactoring commits.

1. ReplicaAsksToLeaveViewMsg,startConsensusProcessConvert, createPrePrepareMessage - convert  raw pointers to smart
2. Convert PrePrepareMsg to smart pointer
3. Convert raw pointers to smart in ReplicaImp::onMessage

* **Testing Done**  
Have not added new tests, we did some heaptrack tests, and masestro tests to check for leaks and stability.

### CODE REVIEW

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 1]
```diff
@@ -5331,15 +5326,13 @@ void ReplicaImp::handleDeferredRequests() {
...
+      auto &msg = deferredRORequests_.front();
       deferredRORequestsMetric_--;
       auto msgHandlerCallback = msgHandlers_->getCallback(msg->type());
       if (msgHandlerCallback) {
-        msgHandlerCallback(std::unique_ptr<MessageBase>(msg));
-      } else {
-        delete msg;
+        msgHandlerCallback(std::unique_ptr<MessageBase>(std::move(msg)));
       }
+      deferredRORequests_.pop_front();
```
Comment (yuliasherman) [inline bftengine/src/bftengine/ReplicaImp.cpp:5335]: In which case do we need this pop?

**Review by eyalrund — APPROVED** [review]

**Comment by yuliasherman** [comment]
> @yuliasherman This is not a review PR, as I've stated its a techincahl merge, approved/reviewed and tested aready. We already reviewed this (everyone had a chance to put there comments and that why we work with a side branch). I only need a technichal approve to merge. I will ask @Efrat1 / @titusen to refer to your comments on the side branch.

I don't remember if I was asked to review the original PR. Please explicitly add me to those reviews in the future.

**Review by yuliasherman — APPROVED** [review]
This PR has to be run for at least 6 hours on Maestro/reserved system to verify it does not add any memory leak and only then merged.

**Comment by glevkovich** [comment]
> > @yuliasherman This is not a review PR, as I've stated its a techincahl merge, approved/reviewed and tested aready. We already reviewed this (everyone had a chance to put there comments and that why we work with a side branch). I only need a technichal approve to merge. I will ask @Efrat1 / @titusen to refer to your comments on the side branch.
> 
> @glevkovich I don't remember if I was asked to review the original PR. Please explicitly add me to those reviews in the future.

@yuliasherman , I've forwarded this PR to others to address your comments on the side branch.

File: bftengine/src/bftengine/ReplicaImp.cpp [hunk 2]
```diff
@@ -622,19 +614,18 @@ bool ReplicaImp::checkSendPrePrepareMsgPrerequisites() {
...
   TimeRecorder scoped_timer(*histograms_.removeDuplicatedRequestsFromQueue);
   // Remove duplicated requests that are result of client retrials from the head of the requestsQueueOfPrimary
-  ClientRequestMsg *first = (!requestsQueueOfPrimary.empty() ? requestsQueueOfPrimary.front() : nullptr);
+  ClientRequestMsg *first = (requestsQueueOfPrimary.empty() ? nullptr : requestsQueueOfPrimary.front().get());
   while (first != nullptr && !clientsManager->canBecomePending(first->clientProxyId(), first->requestSeqNum())) {
     primaryCombinedReqSize -= first->size();
     requestsQueueOfPrimary.pop();
-    delete first;
-    first = (!requestsQueueOfPrimary.empty() ? requestsQueueOfPrimary.front() : nullptr);
+    first = (requestsQueueOfPrimary.empty() ? nullptr : requestsQueueOfPrimary.front().get());
```
[the remark this answers is not in the document]
Comment (titusen) [reply bftengine/src/bftengine/ReplicaImp.cpp:621]: @glevkovich, I've changed due to your request. What's the convention?

[the remark this answers is not in the document]
Comment (glevkovich) [reply bftengine/src/bftengine/ReplicaImp.cpp:621]: @titusen 
There are argument for pros and cons.
It is correct that it's more efficient to use the most expected. On the other hand in terms of readibility i violates other principle (and this time really about style) - don't use double negative when writing code e.g "not empty" or calling a flag isNotActive and writing code like "if !isNotActive".
I think in this case the performance factor is stringer...my fault.

**Pull request merged**